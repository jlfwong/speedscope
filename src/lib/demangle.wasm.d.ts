interface WasmDemangleModule {
  wasm_demangle(mangled: string): string
}

export default function ModuleFactory(options?: unknown): Promise<WasmDemangleModule>
