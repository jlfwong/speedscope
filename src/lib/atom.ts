// This is a very small application state management library.
//
// All it provides is a way of specifying state "atoms" (basically a value with
// a setter and a way to be notified when the value updates), and a single
// preact hook to manage the subscribe/unsubscribe process for you.
//
// At the moment, atoms are intended to be globally defined, but the system
// could easily be adapted to pass down the atoms via context rather than being
// globally available.
//
// To support complex data-types, you can either have simple functions that call
// the public setter, or subclass Atom. So, for example, here are a few
// different ways of making a list atom with convenient helpers:
//
// # Simple functions being passed atom
//
//     const myListAtom = new Atom<number[]>([], "myList")
//     function push<T>(atom: Atom<T[]>, t: T) {
//       const next = [...atom.get(), t]
//       atom.set(next)
//     }
//     function removeLast<T>(atom: Atom<T[]>, t: T) {
//       const next = [...atom.get()]
//       next.pop()
//       atom.set(next)
//     }
//
//
// # Simple functions operating on global atoms
//
//     const myListAtom = new Atom<number[]>([], "myLisT")
//     function push(t: number) {
//       myListAtom.set([...myListAtom.get(), t])
//     }
//     function removeLast() {
//       const next = [...myListAtom.get()]
//       myListAtom.set(next)
//     }
//
//
// # Subclassing
//
//     class ListAtom<T> extends Atom<T[]> {
//        push(t: T) {
//          const next = [...this.state, t]
//          this.set(next)
//        }
//        removeLast() {
//          const next = [...this.state]
//          next.pop()
//          this.set(next)
//        }
//     }
//     const myListAtom = new ListAtom<number>([], "myList")
//
// This library is inspired by https://recoiljs.org/

import {useLayoutEffect, useState} from 'preact/hooks'

type AtomListener = () => void

let AtomDev: {[key: string]: Atom<any>} | null = null
let hotReloadStash: Map<string, any> | null = null

declare const module: any
if (process.env.NODE_ENV === 'development') {
  ;(window as any)['Atom'] = AtomDev = {}

  module.hot.dispose(() => {
    if (AtomDev) {
      hotReloadStash = new Map()
      for (let key in AtomDev) {
        hotReloadStash.set(key, AtomDev[key].get())
      }
    }

    ;(window as any)['Atom_hotReloadStash'] = hotReloadStash
  })

  hotReloadStash = (window as any)['Atom_hotReloadStash'] || null
}

export class Atom<T> {
  private observers: AtomListener[] = []
  constructor(protected state: T, debugKey: string) {
    if (process.env.NODE_ENV === 'development') {
      if (hotReloadStash?.has(debugKey)) {
        // If we have a stored value from a previous hot reload, use that
        // instead of whatever was passed to the constructor.
        this.state = hotReloadStash.get(debugKey)
      }

      if (AtomDev) {
        if (debugKey in AtomDev) {
          console.warn(`[Atom] Multiple atoms tried to register with the key ${debugKey}`)
        }
        AtomDev[debugKey] = this
      }
    }

    // We do the bind here rather than in the definition to facilitate
    // inheritance (we want the value defined on both the prototype and the
    // instance).
    this.set = this.set.bind(this)
    this.get = this.get.bind(this)
  }

  set(t: T) {
    if (this.state === t) {
      // No-op if the value didn't actually change
      return
    }
    this.state = t
    this.observers.forEach(fn => fn())
  }

  get(): T {
    return this.state
  }

  subscribe(listener: AtomListener) {
    this.observers.push(listener)
  }

  unsubscribe(listener: AtomListener) {
    const index = this.observers.indexOf(listener)
    if (index !== -1) {
      this.observers.splice(index, 1)
    }
  }
}

export function useAtom<T>(atom: Atom<T>): T {
  const [value, setValue] = useState(atom.get())

  useLayoutEffect(() => {
    // We need to setValue here because it's possible something has changed the
    // value in the store between the atom.get() call above and layout. In most
    // cases this should no-op.
    setValue(atom.get())

    function listener() {
      setValue(atom.get())
    }

    atom.subscribe(listener)
    return () => {
      atom.unsubscribe(listener)
    }
  }, [atom])

  return value
}
