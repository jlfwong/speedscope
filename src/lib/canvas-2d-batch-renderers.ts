// This file contains a collection of classes which make it easier to perform
// batch rendering of Canvas2D primitives. The advantage of this over just doing
// ctx.beginPath() ... ctx.rect(...) ... ctx.endPath() is that you can construct
// several different batch renderers are the same time, then decide on their
// paint order at the end.
//
// See FlamechartPanZoomView.renderOverlays for an example of how this is used.

export interface TextArgs {
  text: string
  x: number
  y: number
}

export class BatchCanvasTextRenderer {
  private argsBatch: TextArgs[] = []

  text(args: TextArgs) {
    this.argsBatch.push(args)
  }

  fill(ctx: CanvasRenderingContext2D, color: string) {
    if (this.argsBatch.length === 0) return
    ctx.fillStyle = color
    for (let args of this.argsBatch) {
      ctx.fillText(args.text, args.x, args.y)
    }
    this.argsBatch = []
  }
}

export interface RectArgs {
  x: number
  y: number
  w: number
  h: number
}

export class BatchCanvasRectRenderer {
  private argsBatch: RectArgs[] = []

  rect(args: RectArgs) {
    this.argsBatch.push(args)
  }

  private drawPath(ctx: CanvasRenderingContext2D) {
    ctx.beginPath()
    for (let args of this.argsBatch) {
      ctx.rect(args.x, args.y, args.w, args.h)
    }
    ctx.closePath()
    this.argsBatch = []
  }

  fill(ctx: CanvasRenderingContext2D, color: string) {
    if (this.argsBatch.length === 0) return
    ctx.fillStyle = color
    this.drawPath(ctx)
    ctx.fill()
  }

  stroke(ctx: CanvasRenderingContext2D, color: string, lineWidth: number) {
    if (this.argsBatch.length === 0) return
    ctx.strokeStyle = color
    ctx.lineWidth = lineWidth
    this.drawPath(ctx)
    ctx.stroke()
  }
}
