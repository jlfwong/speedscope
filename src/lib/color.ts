import {clamp} from './math'

export class Color {
  constructor(
    readonly r: number = 0,
    readonly g: number = 0,
    readonly b: number = 0,
    readonly a: number = 1,
  ) {}

  static fromLumaChromaHue(L: number, C: number, H: number) {
    // 0 <= L <= 1
    // 0 <= C <= 1
    // 0 <= H <= 360
    // https://en.wikipedia.org/wiki/HSL_and_HSV#From_luma/chroma/hue

    const hPrime = H / 60
    const X = C * (1 - Math.abs(hPrime % 2 - 1))
    const [R1, G1, B1] =
      hPrime < 1
        ? [C, X, 0]
        : hPrime < 2
          ? [X, C, 0]
          : hPrime < 3
            ? [0, C, X]
            : hPrime < 4
              ? [0, X, C]
              : hPrime < 5
                ? [X, 0, C]
                : [C, 0, X]

    const m = L - (0.3 * R1 + 0.59 * G1 + 0.11 * B1)

    return new Color(clamp(R1 + m, 0, 1), clamp(G1 + m, 0, 1), clamp(B1 + m, 0, 1), 1.0)
  }

  toCSS(): string {
    return `rgba(${(255 * this.r).toFixed()}, ${(255 * this.g).toFixed()}, ${(
      255 * this.b
    ).toFixed()}, ${this.a.toFixed(2)})`
  }
}
