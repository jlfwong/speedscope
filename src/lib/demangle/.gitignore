gcc
demangle.wasm.esnext.js
