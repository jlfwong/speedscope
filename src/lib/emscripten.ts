type EmscriptenSymbolMap = Map<string, string>

// This imports symbol maps generated by emscripten using the "--emit-symbol-map" flag.
// It allows you to visualize a profile captured in a release build as long as you also
// have the associated symbol map. To do this, first drop the profile into speedscope
// and then drop the symbol map. After the second drop, the symbols will be remapped to
// their original names.
export function importEmscriptenSymbolMap(contents: string): EmscriptenSymbolMap | null {
  const lines = contents.split('\n')
  if (!lines.length) return null

  // Remove a trailing blank line if there is one
  if (lines[lines.length - 1] === '') lines.pop()
  if (!lines.length) return null

  const map: EmscriptenSymbolMap = new Map()
  const intRegex = /^(\d+):([\$\w-]+)$/
  const idRegex = /^([\$\w]+):([\$\w-]+)$/

  for (const line of lines) {
    // Match lines like "103:__ZN8tinyxml210XMLCommentD0Ev"
    const intMatch = intRegex.exec(line)
    if (intMatch) {
      map.set(`wasm-function[${intMatch[1]}]`, intMatch[2])
      continue
    }

    // Match lines like "u6:__ZN8tinyxml210XMLCommentD0Ev"
    const idMatch = idRegex.exec(line)
    if (idMatch) {
      map.set(idMatch[1], idMatch[2])
      continue
    }

    return null
  }

  return map
}
