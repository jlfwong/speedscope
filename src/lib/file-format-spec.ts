// This file contains types which specify the speedscope file format.

export namespace FileFormat {
  export type Profile = EventedProfile | SampledProfile

  export interface File {
    $schema: 'https://www.speedscope.app/file-format-schema.json'

    // Data shared between profiles
    shared: {
      frames: Frame[]
    }

    // List of profile definitions
    profiles: Profile[]

    // The name of the contained profile group. If omitted, will use the name of
    // the file itself.
    // Added in 0.6.0
    name?: string

    // The index into the `profiles` array that should be displayed upon file
    // load. If omitted, will default to displaying the first profile in the
    // file.
    //
    // Added in 0.6.0
    activeProfileIndex?: number

    // The name of the the program which exported this profile. This isn't
    // consumed but can be helpful for debugging generated data by seeing what
    // was generating it! Recommended format is "name@version". e.g. when the
    // file was exported by speedscope v0.6.0 itself, it will be
    // "speedscope@0.6.0"
    //
    // Added in 0.6.0
    exporter?: string
  }

  export interface Frame {
    name: string
    file?: string
    line?: number
    col?: number
  }

  export enum ProfileType {
    EVENTED = 'evented',
    SAMPLED = 'sampled',
  }

  export interface IProfile {
    // Type of profile. This will future proof the file format to allow many
    // different kinds of profiles to be contained and each type to be part of
    // a discriminated union.
    type: ProfileType
  }

  export interface EventedProfile extends IProfile {
    type: ProfileType.EVENTED

    // Name of the profile. Typically a filename for the source of the profile.
    name: string

    // Unit which all value are specified using in the profile.
    unit: ValueUnit

    // The starting value of the profile. This will typically be a timestamp.
    // All event values will be relative to this startValue.
    startValue: number

    // The final value of the profile. This will typically be a timestamp. This
    // must be greater than or equal to the startValue. This is useful in
    // situations where the recorded profile extends past the end of the recorded
    // events, which may happen if nothing was happening at the end of the
    // profile.
    endValue: number

    // List of events that occured as part of this profile.
    // The "at" field of every event must be in non-decreasing order.
    events: (OpenFrameEvent | CloseFrameEvent)[]
  }

  // List of indices into the frame array
  type SampledStack = number[]

  export interface SampledProfile extends IProfile {
    type: ProfileType.SAMPLED

    // Name of the profile. Typically a filename for the source of the profile.
    name: string

    // Unit which all value are specified using in the profile.
    unit: ValueUnit

    // The starting value of the profile. This will typically be a timestamp.
    // All event values will be relative to this startValue.
    startValue: number

    // The final value of the profile. This will typically be a timestamp. This
    // must be greater than or equal to the startValue. This is useful in
    // situations where the recorded profile extends past the end of the recorded
    // events, which may happen if nothing was happening at the end of the
    // profile.
    endValue: number

    // List of stacks
    samples: SampledStack[]

    // The weight of the sample at the given index. Should have
    // the same length as the samples array.
    weights: number[]
  }

  export type ValueUnit =
    | 'none'
    | 'nanoseconds'
    | 'microseconds'
    | 'milliseconds'
    | 'seconds'
    | 'bytes'

  export enum EventType {
    OPEN_FRAME = 'O',
    CLOSE_FRAME = 'C',
  }

  interface IEvent {
    type: EventType
    at: number
  }

  // Indicates a stack frame opened. Every opened stack frame must have a
  // corresponding close frame event, and the ordering must be balanced.
  interface OpenFrameEvent extends IEvent {
    type: EventType.OPEN_FRAME
    // An index into the frames array in the shared data within the profile
    frame: number
  }

  interface CloseFrameEvent extends IEvent {
    type: EventType.CLOSE_FRAME
    // An index into the frames array in the shared data within the profile
    frame: number
  }
}
