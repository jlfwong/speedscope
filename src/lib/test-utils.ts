import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfileGroup} from '../import'
import {exportProfileGroup, importSpeedscopeProfiles} from './file-format'

interface DumpedProfile {
  name: string
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    name: profile.getName(),
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')

  const profileGroup = await importProfileGroup(path.basename(filepath), input)
  if (profileGroup) {
    expect(profileGroup.name).toMatchSnapshot('profileGroup.name')
    expect(profileGroup.indexToView).toMatchSnapshot('indexToView')

    for (let profile of profileGroup.profiles) {
      expect(dumpProfile(profile)).toMatchSnapshot()
    }
  } else {
    fail('Failed to extract profile')
    return
  }

  const profilesWithoutFilename = await importProfileGroup('unknown', input)
  if (profilesWithoutFilename) {
    expect(profilesWithoutFilename.profiles.length).toEqual(profileGroup.profiles.length)
    profilesWithoutFilename.name = profileGroup.name
    for (let i = 0; i < profileGroup.profiles.length; i++) {
      const a = profileGroup.profiles[i]
      const b = profilesWithoutFilename.profiles[i]
      b.setName(a.getName())
    }
    expect(exportProfileGroup(profileGroup)).toEqual(exportProfileGroup(profilesWithoutFilename))
  } else {
    fail('Failed to extract profile when filename was "unknown"')
    return
  }

  const exported = exportProfileGroup(profileGroup)
  const reimportedGroup = importSpeedscopeProfiles(exported)

  expect(reimportedGroup.name).toEqual(profileGroup.name)
  expect(reimportedGroup.profiles.length).toEqual(profileGroup.profiles.length)

  for (let i = 0; i < profileGroup.profiles.length; i++) {
    const profile = profileGroup.profiles[i]
    const reimported = reimportedGroup.profiles[i]

    expect(reimported.getTotalWeight()).toEqual(profile.getTotalWeight())
    expect(dumpProfile(reimported).stacks.join('\n')).toEqual(
      dumpProfile(profile).stacks.join('\n'),
    )
  }

  const reexported = exportProfileGroup(reimportedGroup)
  expect(exported).toEqual(reexported)
}
