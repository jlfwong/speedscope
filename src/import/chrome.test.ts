import {checkProfileSnapshot} from '../lib/test-utils'

test('importFromChromeCPUProfile', async () => {
  await checkProfileSnapshot('./sample/profiles/Chrome/65/simple.cpuprofile')
})

test('importFromChromeTimeline', async () => {
  await checkProfileSnapshot('./sample/profiles/Chrome/65/simple-timeline.json')
})
