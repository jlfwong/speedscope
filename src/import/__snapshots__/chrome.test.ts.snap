// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importFromChromeCPUProfile 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": 0,
      "file": "",
      "key": "(anonymous)::0:0",
      "line": 0,
      "name": "(anonymous)",
      "selfWeight": 11575,
      "totalWeight": 45920,
    },
    Frame {
      "col": 10,
      "file": "",
      "key": "a::0:10",
      "line": 0,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 34345,
    },
    Frame {
      "col": 10,
      "file": "",
      "key": "b::5:10",
      "line": 5,
      "name": "b",
      "selfWeight": 0,
      "totalWeight": 16788,
    },
    Frame {
      "col": 10,
      "file": "",
      "key": "d::13:10",
      "line": 13,
      "name": "d",
      "selfWeight": 34345,
      "totalWeight": 34345,
    },
    Frame {
      "col": 10,
      "file": "",
      "key": "c::9:10",
      "line": 9,
      "name": "c",
      "selfWeight": 0,
      "totalWeight": 17557,
    },
  ],
  "name": "simple.cpuprofile",
  "stacks": Array [
    "(anonymous) 11.57ms",
    "(anonymous);a;b;d 16.79ms",
    "(anonymous);a;c;d 17.56ms",
  ],
}
`;

exports[`importFromChromeCPUProfile: indexToView 1`] = `0`;

exports[`importFromChromeCPUProfile: profileGroup.name 1`] = `"simple.cpuprofile"`;

exports[`importFromChromeTimeline 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": -1,
      "file": "",
      "key": "(program)::-1:-1",
      "line": -1,
      "name": "(program)",
      "selfWeight": 12244,
      "totalWeight": 12244,
    },
    Frame {
      "col": 10,
      "file": "file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "a:file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:0:10",
      "line": 0,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 33184,
    },
    Frame {
      "col": 10,
      "file": "file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "c:file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:13:10",
      "line": 13,
      "name": "c",
      "selfWeight": 16197,
      "totalWeight": 16197,
    },
    Frame {
      "col": 10,
      "file": "file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "b:file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:7:10",
      "line": 7,
      "name": "b",
      "selfWeight": 16987,
      "totalWeight": 16987,
    },
  ],
  "name": "simple-timeline.json",
  "stacks": Array [
    "(program) 10.16ms",
    " 1.99ms",
    "(program) 926.00µs",
    " 85.95ms",
    "(program) 133.00µs",
    " 8.40ms",
    "(program) 134.00µs",
    "a;c 257.00µs",
    "a;b 534.00µs",
    "a;c 513.00µs",
    "a;b 513.00µs",
    "a;c 257.00µs",
    "a;b 518.00µs",
    "a;c 514.00µs",
    "a;b 261.00µs",
    "a;c 517.00µs",
    "a;b 258.00µs",
    "a;c 145.00µs",
    "a;b 134.00µs",
    "a;c 292.00µs",
    "a;b 272.00µs",
    "a;c 129.00µs",
    "a;b 145.00µs",
    "a;c 265.00µs",
    "a;b 142.00µs",
    "a;c 141.00µs",
    "a;b 139.00µs",
    "a;c 397.00µs",
    "a;b 126.00µs",
    "a;c 131.00µs",
    "a;b 126.00µs",
    "a;c 521.00µs",
    "a;b 128.00µs",
    "a;c 261.00µs",
    "a;b 127.00µs",
    "a;c 130.00µs",
    "a;b 128.00µs",
    "a;c 136.00µs",
    "a;b 272.00µs",
    "a;c 146.00µs",
    "a;b 661.00µs",
    "a;c 113.00µs",
    "a;b 135.00µs",
    "a;c 585.00µs",
    "a;b 294.00µs",
    "a;c 274.00µs",
    "a;b 129.00µs",
    "a;c 127.00µs",
    "a;b 128.00µs",
    "a;c 247.00µs",
    "a;b 526.00µs",
    "a;c 125.00µs",
    "a;b 129.00µs",
    "a;c 131.00µs",
    "a;b 253.00µs",
    "a;c 135.00µs",
    "a;b 123.00µs",
    "a;c 128.00µs",
    "a;b 257.00µs",
    "a;c 137.00µs",
    "a;b 256.00µs",
    "a;c 411.00µs",
    "a;b 892.00µs",
    "a;c 146.00µs",
    "a;b 148.00µs",
    "a;c 309.00µs",
    "a;b 140.00µs",
    "a;c 151.00µs",
    "a;b 458.00µs",
    "a;c 156.00µs",
    "a;b 170.00µs",
    "a;c 139.00µs",
    "a;b 290.00µs",
    "a;c 115.00µs",
    "a;b 281.00µs",
    "a;c 270.00µs",
    "a;b 283.00µs",
    "a;c 144.00µs",
    "a;b 146.00µs",
    "a;c 410.00µs",
    "a;b 146.00µs",
    "a;c 147.00µs",
    "a;b 312.00µs",
    "a;c 146.00µs",
    "a;b 907.00µs",
    "a;c 231.00µs",
    "a;b 138.00µs",
    "a;c 280.00µs",
    "a;b 255.00µs",
    "a;c 257.00µs",
    "a;b 130.00µs",
    "a;c 255.00µs",
    "a;b 128.00µs",
    "a;c 526.00µs",
    "a;b 128.00µs",
    "a;c 384.00µs",
    "a;b 398.00µs",
    "a;c 260.00µs",
    "a;b 664.00µs",
    "a;c 150.00µs",
    "a;b 147.00µs",
    "a;c 156.00µs",
    "a;b 137.00µs",
    "a;c 141.00µs",
    "a;b 140.00µs",
    "a;c 141.00µs",
    "a;b 141.00µs",
    "a;c 140.00µs",
    "a;b 140.00µs",
    "a;c 139.00µs",
    "a;b 129.00µs",
    "a;c 269.00µs",
    "a;b 293.00µs",
    "a;c 148.00µs",
    "a;b 293.00µs",
    "a;c 292.00µs",
    "a;b 296.00µs",
    "a;c 147.00µs",
    "a;b 140.00µs",
    "a;c 139.00µs",
    "a;b 139.00µs",
    "a;c 511.00µs",
    "a;b 515.00µs",
    "a;c 386.00µs",
    "a;b 514.00µs",
    "a;c 272.00µs",
    "a;b 242.00µs",
    "a;c 128.00µs",
    "a;b 134.00µs",
    "a;c 112.00µs",
    "a;b 288.00µs",
    "a;c 451.00µs",
    "a;b 138.00µs",
    "a;c 292.00µs",
    "a;b 145.00µs",
    "a;c 259.00µs",
    "a;b 148.00µs",
    "a;c 293.00µs",
    "a;b 140.00µs",
    "a;c 140.00µs",
    " 61.61ms",
    "(program) 118.00µs",
    " 134.00µs",
    "(program) 128.00µs",
    " 16.40ms",
    "(program) 125.00µs",
    " 16.64ms",
    "(program) 129.00µs",
    " 16.72ms",
    "(program) 130.00µs",
    " 294.15ms",
    "(program) 129.00µs",
    " 7.68ms",
    "(program) 131.00µs",
  ],
}
`;

exports[`importFromChromeTimeline: indexToView 1`] = `0`;

exports[`importFromChromeTimeline: profileGroup.name 1`] = `"simple-timeline.json"`;
